### PR TITLE
Feature/socket server

### PR DESCRIPTION
**Pull request by creisle**
- Socket server back-end for the wizzards app

Outstanding Features
- detect game end
- leave game

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -10,15 +10,22 @@
...
     "react-dom": "^16.13.1",
     "react-scripts": "3.4.1",
+    "shortid": "^2.2.15",
+    "socket.io": "^2.3.0",
     "typescript": "^3.7.5"
   },
   "scripts": {
     "start": "react-scripts start",
     "build": "react-scripts build",
-    "test": "react-scripts test",
```
Comment (areisle) [inline package.json:35]: why did you remove 'test'?

File: server/db/round.js [hunk 2]
```diff
@@ -0,0 +1,420 @@
...
+    const cardsUsed = (round + 1) * players.length;
+    const promises = [];
+
+    let trumpSuit = 'jester';
+    if (cardsUsed < TOTAL_CARDS) {
+        // select a trump card
+        trumpSuit = deck[cardsUsed].suit;
+    }
+    if (trumpSuit === 'wizards') {
+        // TODO: let the player decide, currently pick random
```
Comment (areisle) [inline server/db/round.js:364]: can you add the todos to the docstrings?

File: server/index.js [hunk 3]
```diff
@@ -0,0 +1,196 @@
...
+            try {
+                const [, round, trick] = await Promise.all([
+                    db.playerExists(redis, gameId, playerId),
+                    db.getCurrentRound(redis, gameId),
+                    db.getCurrentTrick(redis, gameId),
+                ]);
+                const { trickComplete, trickWinner, roundComplete, newLeadSuit } = await db.playCard(
+                    redis, gameId, playerId, cardSuit, cardValue
+                );
+                // end of round? or start next trick
```
Comment (areisle) [inline server/index.js]: leftover?

File: server/db/index.js [hunk 4]
```diff
@@ -0,0 +1,129 @@
...
+const connect = async () => {
+    const redis = new Redis();
+    try {
+        await redis.connect();
+    } catch (err) { }
+    return redis;
+};
+
+
+const getGameState = async (redis, gameId) => {
```
Comment (areisle) [inline server/db/index.js]: should this be in the game file?

File: server/db/index.js [hunk 5]
```diff
@@ -0,0 +1,129 @@
...
+        });
+
+        for (const trickWinner of trickWinners[round]) {
+            scores[round][trickWinner].taken += 1;
+        }
+    });
+
+    const activePlayer = await whosTurnIsIt(redis, gameId);
+
+    return {
```
Comment (areisle) [inline server/db/index.js:76]: this should also return the players cards, the trickLeader, and the cards played so far for the trick

File: package.json (hunk 1 above)
Comment (creisle) [reply package.json:35]: I just replaced it with test:server until we have tests for the client. I thought it would be confusing to have it default to running the server tests since all the other defaults are for the client half

File: server/index.js (hunk 3 above)
Comment (creisle) [reply server/index.js]: yup, will remove

File: server/db/index.js (hunk 4 above)
Comment (creisle) [reply server/db/index.js]: no b/c it would cause some circular imports

File: server/db/index.js (hunk 5 above)
Comment (creisle) [reply server/db/index.js:76]: would adding something like below work?

```
{
  trick: {
    number: 1, // previous trick
    cards: [...],
    leader: 'name'
  }
}
```

Comment (areisle) [reply server/db/index.js:76]: I would rather have the cards as `{ username: card }` as that's how I'm storing them in the client.

I'm not sure what `// previous trick` is about?

Comment (creisle) [reply server/db/index.js:76]: sorry, that should have been "previously trick" was suggesting moving the trick number from ".trick" to ".trick.number"

Comment (creisle) [reply server/db/index.js:76]: > I would rather have the cards as `{ username: card }` as that's how I'm storing them in the client.

you'd don't want the card order?

Comment (areisle) [reply server/db/index.js:76]: if it makes no difference on your end, can they just be 'trickNumber', 'trickCards', and 'trickLeader'? otherwise I just have to transform them all to that

Comment (areisle) [reply server/db/index.js:76]: > > I would rather have the cards as `{ username: card }` as that's how I'm storing them in the client.
> 
> you'd don't want the card order?

no thats much more difficult to work with

**Review by areisle — APPROVED** [review]
